Added overLogAction to HasLog typeclass

sphaso opened the pull request:
Resolves #85 

I'm not 100% certain that I did what you meant with:
> And add default implementation to setLogAction function.

cronokirby commented:
@sphaso You can implement `setLogAction` via `overLogAction`, so it'd be good to have a default implementation in the typeclass that way, allowing users to specialise it if they have a more efficient solution.

I believe that's what was being requested in the issue

ChShersh commented on co-log-core/src/Colog/Core/Class.hs:
```diff
@@ -18,11 +17,9 @@ TODO: laws
 class HasLog env msg m where
     getLogAction :: env -> LogAction m msg
     setLogAction :: LogAction m msg -> env -> env
+    overLogAction :: (LogAction m msg -> LogAction m msg) -> env -> env
```
Yes, you should add default implementations of `overLogAction` and `setLogAction`

cronokirby replied:
Shouldn't we just define one in terms of the other?
Is there a sensible default implementation for *both* at once?

ChShersh replied:
By default I mean no `default` keyword, just like for `Eq` typeclass where `==` is implemented using `/=` and `/=` is implemented using `==`

ChShersh commented on co-log-core/src/Colog/Core/Class.hs:
```diff
@@ -18,11 +17,9 @@ TODO: laws
 class HasLog env msg m where
     getLogAction :: env -> LogAction m msg
```
Also it's a good idea to add `{-# MINIMAL #-}` pragma to this typeclass.

ChShersh commented on co-log-core/src/Colog/Core/Class.hs:
```diff
@@ -18,11 +17,9 @@ TODO: laws
-overLogAction :: HasLog env msg m => (LogAction m msg -> LogAction m msg) -> env -> env
-overLogAction over env = setLogAction (over $ getLogAction env) env
-{-# INLINE overLogAction #-}
+    overLogAction over = over
```
Looks like this is also `id`
```suggestion
    overLogAction = id
```

sphaso commented:
@cronokirby @ChShersh    
- added change to CHANGELOG
- added MINIMAL pragma
- refactored `overLogAction`
- added (very naive) default implementations for `setLogAction` and `overLogAction`

cronokirby commented on co-log-core/src/Colog/Core/Class.hs:
```diff
@@ -18,11 +17,12 @@ TODO: laws
 class HasLog env msg m where
     getLogAction :: env -> LogAction m msg
     setLogAction :: LogAction m msg -> env -> env
+    setLogAction = seq
```
Why `seq` instead of `flip const`? Why are we forcing evaluation of the LogAction?

sphaso replied:
sorry, I'm not sure I'm following, aren't `seq` and `flip const` equal?
```
>> :t seq
seq :: a -> b -> b
>> :t flip const
flip const :: b -> c -> c
```

ChShersh replied:
`setLogAction` here should be implemented using `overLogAction`.

Also, don't forget to add `{-# INLINE #-}` pragmas to the default implementations 🙂

cronokirby commented on co-log-core/src/Colog/Core/Class.hs:
```diff
@@ -18,11 +17,12 @@ TODO: laws
-overLogAction :: HasLog env msg m => (LogAction m msg -> LogAction m msg) -> env -> env
-overLogAction over env = setLogAction (over $ getLogAction env) env
-{-# INLINE overLogAction #-}
+    setLogAction = overLogAction . const
```
IMO the default implementation of setLogAction should be this in the typeclass, and then we don't even have to define it in this trivial case here

sphaso commented:
@ChShersh from the looks of it, it might make more sense to have a default implementation of `setLogAction` and only require the user to set `overLogAction`. I think @cronokirby has a point that they could be implemented in terms of each other as a default.

ChShersh commented on co-log-core/CHANGELOG.md:
```diff
@@ -3,6 +3,11 @@
+## 0.1.2
+
+* [#85](https://github.com/kowainik/co-log/issues/85):
+  Move `overLogAction` to `HasLog` typeclass
```
I wonder whether it's a breaking change and should go to `0.2.0` or not 🤔

ChShersh commented on co-log-core/src/Colog/Core/Class.hs:
```diff
@@ -18,11 +17,12 @@ TODO: laws
     setLogAction :: LogAction m msg -> env -> env
+    setLogAction = seq
+    overLogAction :: (LogAction m msg -> LogAction m msg) -> env -> env
+    overLogAction = seq
```
Also, not `seq`, you should use `getLogAction` and `setLogAction` to implement `overLogAction`

ChShersh commented on co-log-core/src/Colog/Core/Class.hs:
```diff
@@ -18,11 +17,12 @@ TODO: laws
+    setLogAction = seq
+    overLogAction :: (LogAction m msg -> LogAction m msg) -> env -> env
+    overLogAction = seq
+    {-# MINIMAL getLogAction #-}
```
Minimal complete set actually should be:

```haskell
{-# MINIMAL getLogAction, (setLogAction | overLogAction) #-}
```

sphaso commented:
@cronokirby @ChShersh hopefully I didn't make any egregious mistake this time :) thanks for your review as always!

ChShersh commented on co-log-core/src/Colog/Core/Class.hs:
```diff
@@ -18,11 +17,13 @@ TODO: laws
-overLogAction :: HasLog env msg m => (LogAction m msg -> LogAction m msg) -> env -> env
-overLogAction over env = setLogAction (over $ getLogAction env) env
-{-# INLINE overLogAction #-}
+    overLogAction = id
```
Let's also implement `setLogAction` function. If you don't implement methods with default definitions from typeclass, then the default implementation of those methods is used. If you implement methods manually, you have better control over the implementation and thus you can have better performance. Also, let's add `{-# INLINE #-}` pragmas to all three functions in this instance to have even more guarantees for performance.

ChShersh commented on co-log-core/src/Colog/Core/Class.hs:
```diff
@@ -18,11 +17,13 @@ TODO: laws
+    overLogAction :: (LogAction m msg -> LogAction m msg) -> env -> env
+    overLogAction f env = setLogAction (f $ getLogAction env) env
+    {-# INLINE overLogAction #-}
+    {-# MINIMAL getLogAction, (setLogAction | overLogAction) #-}
```
Can you, please, run haddock command for `co-log-core` and post screenshot of how it looks like? I hope that syntax is correct and there should be no problems with rendering it... Also, it's usually a good practice to put `{-# MINIMAL #-}` pragma before all methods inside typeclass definition.

sphaso commented:
@ChShersh 
- added INLINE everywhere
- moved MINIMAL on top
- implemented `setLogAction` in the instance

This is the [Haddock output in HTML](https://imgur.com/a/gOiPGkI)

ChShersh commented on co-log-core/src/Colog/Core/Class.hs:
```diff
@@ -16,13 +15,19 @@ import Colog.Core.Action (LogAction)
-overLogAction over env = setLogAction (over $ getLogAction env) env
-{-# INLINE overLogAction #-}
+    {-# INLINE setLogAction #-}
+    setLogAction = flip const
```
Looks like a minor error in the implementation
```suggestion
    setLogAction = const
```

ChShersh commented:
@sphaso Let's keep it minor version bump up. I tried to think about this more and looks like it's not a breaking change. `overLogAction` has default definition, so existing instances won't break. Import `Colog (HasLog (..), overLogAction)` doesn't result into error, so it's still a valid Haskell code.
